### PR TITLE
New selection box implementation for component groups

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

[Task link](https://www.pivotaltracker.com/story/show/182194574).
[ci no changelog needed]

This PR implements a new selection box that will replace an old (not really working) one in the component browser. The old selection box wasn't working well with the headers of the component groups, so we were forced to make a much harder implementation.

The new implementation duplicates some visual components and places them in a separate layer. Then, a rectangular mask cuts off everything that is not "selected". This way:
- We have more control over what the selected entries should look like.
- We can easily support the multi-layer structure of the component groups with headers.
- We avoid problems with nested masks that our renderer doesn't support at the moment.

To be more precise, we duplicate the following:
- Background of the component group becomes the "fill" of the selection.
- Entries text and icons - we can alter them easily.
- Header background and header text. By placing them in separate scene layers we ensure correct rendering order.

https://user-images.githubusercontent.com/6566674/173657899-1067f538-4329-44f9-9dc2-78c8a4708b5a.mp4

### Important Notes

- This PR implements the base of our future selection mechanism, selecting entries with a mouse and keyboard still has several issues that will be fixed in the future tasks.
- The scrolling behavior will also be improved in future tasks. Right we only restrict the selection box position so that it never leaves the borders of the component group.
- I added a new function to `add` shapes to new layers in a non-exclusive way (we had only `add_exclusive`) before. I have no idea how we didn't use this feature before even though we mention it a lot in the docs.
- The demo scene restricts the position of the selection box for one-column component groups but does not for the wide component group. 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 1]
```diff
@@ -127,18 +131,33 @@ pub struct View {
     icon_strong_color: frp::Sampler<color::Rgba>,
     icon_weak_color:   frp::Sampler<color::Rgba>,
     label:             GlyphHighlightedLabel,
+    selected_label:    GlyphHighlightedLabel,
+    selected_layer:    Rc<Option<WeakLayer>>,
+    label_layer:       Rc<RefCell<WeakLayer>>,
 }
 
-impl list_view::Entry for View {
+impl Entry for View {
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/entry.rs]: Why this change? I prefer to see that the trait is a `list_view` thing.

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 2]
```diff
@@ -78,6 +80,7 @@ impl From<&str> for Model {
 #[derive(Clone, CloneRef, Debug)]
 pub struct Params {
     pub colors: Colors,
+    pub layer:  Rc<Option<WeakLayer>>,
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/entry.rs]: I don't see what is that layer, and what it is for; generally, the entries should inherit layers from their parent. And for label layer, there is a `set_label_layer` method.

**Comment by wdanilo** [comment]
First visual feedback:
1. The selection animation is too slow. Please make it 2x faster.
2. Could you show on a video what happens to the selection when you scroll the list so the selection goes out of scope (on both end), please?

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 3]
```diff
@@ -127,18 +131,33 @@ pub struct View {
...
         let label = GlyphHighlightedLabel::new(app, style_prefix, &());
+        let selected_label = GlyphHighlightedLabel::new(app, style_prefix, &());
         display_object.add_child(&label);
+        display_object.add_child(&selected_label);
+
+        if let Some(selected_layer) = &**layer {
+            if let Some(layer) = selected_layer.upgrade() {
+                selected_label.set_label_layer(&layer);
+            }
+        }
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/entry.rs]: IMO, if there is no selected_layer, it would make sense to not display selected_label at all. And print a warning/error log - I don't think the dropped layer is something we expect.

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 4]
```diff
@@ -189,6 +221,18 @@ impl list_view::Entry for View {
             shape.weak_color.set(self.icon_weak_color.value().into());
             shape.set_position_x(icon::SIZE / 2.0);
             self.display_object.add_child(&shape);
+            if let Some(label_layer) = self.label_layer.borrow().upgrade() {
+                label_layer.add_exclusive(&shape);
+            } else {
+                error!(self.logger, "Label layer of the component group entry was dropped.");
+            }
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/entry.rs]: I think we should not keep icons on the label layer; the label layer is, well, for labels, and is actually a workaround for the inability of text to be sorted with other shapes in one layer. 

If we do it for having icons above selection, we shall use "above" clause in their definition in app/gui/view/component-browser/component-group/src/icon.rs.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 5]
```diff
@@ -15,6 +15,25 @@
...
+//!
+//! So instead we duplicate component group entries in a special `selection` scene layer (or
+//! rather, in a multiple scene layers to ensure render ordering) and use [layers masking][mask]
+//! to cut off everything except the selection shape. We duplicate the following:
+//! - Entry text and icon (see [entry][] module).
+//! - A background of the component group [selection_background][].
+//! - (for component groups with header) A background of the header [selection_header_background][].
+//! - (for component groups with header) Header text.
+//!
+//! [mask]: ensogl::display::scene::layer::Layer#masking-layers-with-arbitrary-shapes
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: I would add to this nice doc section an info that this allows us also to have text/icon in different color when highlighted, so it looks good during selection box animation.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 6]
```diff
@@ -91,6 +107,40 @@ const HEADER_SHADOW_PEAK: f32 = list_view::entry::HEIGHT / 2.0;
...
+
+    ensogl::define_shape_system! {
+        pointer_events = false;
+        (color:Vector4) {
+            let color = Var::<color::Rgba>::from(color);
+            Plane().fill(color).into()
+        }
+    }
+}
+
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: Cannot we just re-use the `background` shape with different color and at different layer?

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 7]
```diff
@@ -108,6 +159,7 @@ pub mod background {
 }
 
 
+
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: AFAIR here should be only two empty lines.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 8]
```diff
@@ -434,55 +511,82 @@ impl component::Frp<Model> for Frp {
...
+    selection: LayersInner,
+}
+
+impl Layers {
+    /// Constructor.
+    pub fn new(
+        logger: &Logger,
+        normal_parent_layer: &Layer,
+        selected_parent_layer: &Layer,
+    ) -> Self {
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: Someone who just spots this constructor may wonder, what is `normal_parent_layer` and `selected_parent_layer` - add at least reference to the method's docs.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 9]
```diff
@@ -434,55 +511,82 @@ impl component::Frp<Model> for Frp {
...
+
+        Self { normal, selection: selected }
+    }
+}
+
 /// A set of scene layers shared by every component group.
 ///
 /// A component group consists of a several shapes with a strict rendering order. The order of the
 /// fields of this struct represents the rendering order of layers, with `background` being the
 /// bottom-most and `header_text` being the top-most.
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: I would put this extended explanation to published `Layer` structure, not LayersInner.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 10]
```diff
@@ -434,55 +511,82 @@ impl component::Frp<Model> for Frp {
...
+    header_text: Layer,
 }
 
-impl Layers {
+impl LayersInner {
     /// Constructor.
     ///
-    /// A `camera` will be used to render all layers. Layers will be attached to a `parent_layer` as
+    /// Layers will be attached to a `parent_layer` as
     /// sublayers.
```
Comment (farmaazon) [inline app/gui/view/component-browser/component-group/src/lib.rs]: Unnecessary line wrapping.

File: app/gui/view/debug_scene/component-group/src/lib.rs [hunk 11]
```diff
@@ -240,31 +244,33 @@ fn init(app: &Application) {
...
-    // [`transparent_circle`].
-    {
-        let transparent_circle = transparent_circle::View::new(&app.logger);
-        transparent_circle.size.set(Vector2(150.0, 150.0));
-        transparent_circle.set_position_xy(Vector2(200.0, -150.0));
-        scroll_area.content().add_child(&transparent_circle);
-        std::mem::forget(transparent_circle);
-    }
+    // This is a workaround for a bug - without this transparent shape the content of the scroll
+    // area is invisible.
```
Comment (farmaazon) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: Why the issue number was removed?

File: lib/rust/ensogl/core/src/display/scene.rs [hunk 12]
```diff
@@ -563,6 +563,8 @@ pub struct HardcodedLayers {
     pub tooltip_text:       Layer,
     pub cursor:             Layer,
     pub mask:               Layer,
+    pub selection:          Layer,
+    pub selection_mask:     Layer,
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/display/scene.rs]: I think those could also be sub-layers of the "node_searcher" layer (or whenever the component groups are attached) - siblings of the scroll area's sub-layers (so they should not inherit mask). 

If you find it too complicated, the can skip it now (the hard-coded layers should be reworked anyway, to make better use of the sublayer mechanism), but add a comment explaining for what those layers are.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs]: This was a mistake, fixed

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 2 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs]: Renamed to `selection_layer`, I forgot to push that change to the branch. Thi slayer is needed to display selected entry text.

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs]: Fixed while refactoring

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs]: The label layer is no longer present in the code, it was removed while refactoring. Normal icon inherits the layer of the display object, selected icon uses the `selection_layer`.

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 5 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Added

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 6 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Actually yes, we can, they are almost identical. Done.

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 8 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Added docs, slightly renamed the arguments.

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 9 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Done

File: app/gui/view/debug_scene/component-group/src/lib.rs (hunk 11 above)
Comment (vitvakatu) [reply app/gui/view/debug_scene/component-group/src/lib.rs]: Merge conflict, fixed

File: lib/rust/ensogl/core/src/display/scene.rs (hunk 12 above)
Comment (vitvakatu) [reply lib/rust/ensogl/core/src/display/scene.rs]: Removed these layer from hardcoded layers and added them to the demo scene instead.

**Comment by vitvakatu** [comment]
@wdanilo 
> 1. The selection animation is too slow. Please make it 2x faster.
    2. Could you show on a video what happens to the selection when you scroll the list so the selection goes out of scope (on both end), please?

1. Increased its speed by changing `spring` param (also exposed API to change it for ensogl::Animation)
2. Updated the screencast. For now, we just don't allow the selection to move beyond the scroll area.

@farmaazon all comments addressed.

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 3 above)
Comment (farmaazon) [reply app/gui/view/component-browser/component-group/src/entry.rs]: The comment about not displaying selected_label when there is no label seems unaddressed. The `display_object.add_child(&selected_label);` could be just inside if.

**Review by farmaazon — APPROVED** [review]
One of my comments in the previous PR seems unaddressed: the one about not displaying selected_label when there is no selection layer.

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 13]
```diff
@@ -105,14 +108,29 @@ impl Default for Params {
...
+    fn set_strong_color(&self, color: color::Rgba) {
+        if let Some(shape) = &self.shape {
+            shape.strong_color.set(color.into());
+        }
+    }
+
+    fn set_weak_color(&self, color: color::Rgba) {
+        if let Some(shape) = &self.shape {
+            shape.weak_color.set(color.into());
+        }
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/entry.rs:138]: "strong" and "weak" color sounds strange. Let's name it "active_color" and "inactive_color" instead.

File: app/gui/view/component-browser/component-group/src/entry.rs [hunk 14]
```diff
@@ -123,22 +141,60 @@ pub struct View {
...
     max_width_px:      frp::Source<f32>,
     icon_strong_color: frp::Sampler<color::Rgba>,
     icon_weak_color:   frp::Sampler<color::Rgba>,
     label:             GlyphHighlightedLabel,
+    selected_label:    GlyphHighlightedLabel,
+    selection_layer:   Rc<Option<WeakLayer>>,
+}
+
+impl View {
+    /// Update an icon shape (create it if necessary), update its color, add it to `layer` if
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/entry.rs]: I know it doesnt work in the generated doc in Rust, but I feel its better to write ``` [`layer`] ``` instead of ``` `layer` ```, becuse you are trying to refer to something, not just create inline code snippet.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 15]
```diff
@@ -15,6 +15,30 @@
...
+//! component group uses 4 different [scene layers][Layer] to render its header correctly, and we
+//! want selection be above the header background and below the text entries at the same time, which
+//! is not possible.
+//!
+//! So instead we duplicate component group entries in a special `selection` scene layer (or
+//! rather, in a multiple scene layers to ensure render ordering) and use [layers masking][mask]
+//! to cut off everything except the selection shape. We duplicate the following:
+//! - Entry text and icon (see [entry][] module).
+//! - A background of the component group [background][].
+//! - (for component groups with header) A background of the header [selection_header_background][].
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/lib.rs:33]: Please use some grammar-fixing tool. There is a lot of places with missing commas and small grammar issues. Apply to all doc comments, please.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 16]
```diff
@@ -91,6 +112,26 @@ const HEADER_SHADOW_PEAK: f32 = list_view::entry::HEIGHT / 2.0;
 // === Shapes Definitions ===
 // ==========================
 
+// === Selection ===
+
+/// A shape of selection box. It is used as a mask to show only specific parts of the selection
+/// layers. See module-level documentation.
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/lib.rs]: `See module-level documentation.` -> `See module-level documentation to learn more.`. Please apply to other docs as well.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 17]
```diff
@@ -334,7 +394,8 @@ impl component::Frp<Model> for Frp {
 
         // === Colors ===
         let colors = Colors::from_main_color(network, style, &input.set_color, &input.set_dimmed);
-        let params = entry::Params { colors: colors.clone_ref() };
+        let params =
+            entry::Params { colors: colors.clone_ref(), selection_layer: default() };
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/lib.rs]: We prefer 2 expressions with 2 variables than one multiline expression. Please, refactor `colors` to a separate variable, and make it one-liner.

File: app/gui/view/component-browser/component-group/src/lib.rs [hunk 18]
```diff
@@ -436,53 +503,80 @@ impl component::Frp<Model> for Frp {
...
-        let background = layer::Layer::new_with_cam(logger.clone_ref(), camera);
-        let text = layer::Layer::new_with_cam(logger.clone_ref(), camera);
-        let header = layer::Layer::new_with_cam(logger.clone_ref(), camera);
-        let header_text = layer::Layer::new_with_cam(logger.clone_ref(), camera);
+    /// `normal` layers are assigned as sublayers of the `normal_parent`, while `selection`
+    /// layers are assigned to the `selected_parent`.
+    pub fn new(logger: &Logger, normal_parent: &Layer, selected_parent: &Layer) -> Self {
+        let normal = LayersInner::new(logger, normal_parent);
+        let selection = LayersInner::new(logger, selected_parent);
+
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/lib.rs]: why this newline?

File: app/gui/view/component-browser/component-group/src/wide.rs [hunk 19]
```diff
@@ -248,13 +252,16 @@ impl<const COLUMNS: usize> component::Frp<Model<COLUMNS>> for Frp {
...
+                    f!((_, size, padding) column.resize_and_place(*size, *padding))
+                );
             }
             frp::extend! { network
                 out.is_mouse_over <+ is_mouse_over;
             }
 
-            let params = entry::Params { colors: colors.clone_ref() };
+            let params =
+                entry::Params { colors: colors.clone_ref(), selection_layer: default() };
```
Comment (wdanilo) [inline app/gui/view/component-browser/component-group/src/wide.rs]: multi-line exprs. Please apply to other places in the code.

File: app/gui/view/debug_scene/component-group/src/lib.rs [hunk 20]
```diff
@@ -40,6 +42,11 @@ use list_view::entry::AnyModelProvider;
 // =================
 
 const COMPONENT_GROUP_COLOR: color::Rgba = color::Rgba::new(0.527, 0.554, 0.18, 1.0);
+/// The selection animation is faster than default one because of the increased spring force.
+const SELECTION_ANIMATION_SPRING_FORCE: f32 = 30_000.0;
```
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: Leave here a multiplier intead. So if we change the spring force globally thsi will update automatically.

File: app/gui/view/debug_scene/component-group/src/lib.rs [hunk 21]
```diff
@@ -343,16 +369,63 @@ fn init(app: &Application) {
...
+    selection.size.set(Vector2(150.0, list_view::entry::HEIGHT));
+    selection_mask.add_exclusive(&selection);
+    app.display.add_child(&selection);
+
+    let selection_animation = Animation::<Vector2>::new(&network);
+    selection_animation.set_spring.emit(SELECTION_ANIMATION_SPRING_FORCE);
+    /// This is an example code to position the selection box on scene.
+    /// `multiview.selection_position_target` returns a group-local position, so we transform it
+    /// to global one and then restrict the y coordinate so that the selection box would not go
+    /// below the scroll area bottom.
```
Comment (wdanilo) [inline app/gui/view/debug_scene/component-group/src/lib.rs]: It's really hard to read these sentences. Please make them shorter and run them trough a grammar checking tool. Gramarly is cool. If you need pro license, Ola will buy you one.

**Review by wdanilo — APPROVED** [review]

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs]: I confused `selected_label` with `selected_icon`, fixed it. However, we still create the label anyway, because making it optional makes code a lot uglier due to multiple `if let` statements and the mutable variable in the constructor.

File: app/gui/view/component-browser/component-group/src/entry.rs (hunk 13 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/entry.rs:138]: They are not `active` or `inactive`, they are used by icons with multiple colored parts involved. This is described in the docs of `icon::Any` (this shape variable), I've also added it to the docs of the `Colors` struct.

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 15 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs:33]: Done

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 16 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Funny, I had your variant in the first version of the code, and then I thought it was too long.

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 17 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Done

File: app/gui/view/component-browser/component-group/src/lib.rs (hunk 18 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/lib.rs]: Removed

File: app/gui/view/component-browser/component-group/src/wide.rs (hunk 19 above)
Comment (vitvakatu) [reply app/gui/view/component-browser/component-group/src/wide.rs]: Done. There are a few places with multiline expressions in the code, but it's impossible to change it without making the code *way* uglier.

File: app/gui/view/debug_scene/component-group/src/lib.rs (hunk 20 above)
Comment (vitvakatu) [reply app/gui/view/debug_scene/component-group/src/lib.rs]: Done

File: app/gui/view/debug_scene/component-group/src/lib.rs (hunk 21 above)
Comment (vitvakatu) [reply app/gui/view/debug_scene/component-group/src/lib.rs]: Done.

**Comment by vitvakatu** [comment]
QA acceptance passed, merging it now.

**Review by MichaelMauderer — APPROVED** [review]
